docs/docs/reference/gen_notebooks/hallucination.md

```diff
@@ -0,0 +1,450 @@
+    "vectara/hallucination_evaluation_model", trust_remote_code=True
+)
+model.eval()  # Set the model to evaluation mode
+
+```
+
+## 🔑 Initialize OpenAI Client and Embedding Model
+
+Create an OpenAI client instance for API calls and set up the embedding model.
+
+```python
+# Initialize OpenAI client
+client = OpenAI(
+    api_key=os.environ.get("OPENAI_API_KEY"),
+)
+
+# Set up embedding model
+embedding_model = OpenAIEmbedding(model="text-embedding-ada-002")
+
+
+```
+
+## 📥 Download and Load Documents
+
+We'll download a PDF document from a URL and create an index using LlamaIndex. Please note taht this can be your own vector database with your data indexed for your RAG Chatbot.
+
+```python
+
+# Download the PDF from a URL
+pdf_url = "https://arxiv.org/pdf/2408.13296v1.pdf"  # Replace with your PDF URL
+pdf_filename = "document.pdf"
+
+response = requests.get(pdf_url)
+with open(pdf_filename, 'wb') as f:
+    f.write(response.content)
+
+# Load the documents from the PDF
+documents = SimpleDirectoryReader(input_dir='.', required_exts=['.pdf']).load_data()
+
+# Create the index from the documents
+index = VectorStoreIndex.from_documents(documents, embed_model=embedding_model)
+
+```
+
+
+## 🔎 Create Query Engine
+
+Set up the query engine with a limit on the number of retrieved documents.
+
+```python
+query_engine = index.as_query_engine(similarity_top_k=3)
+```
+
+## 🛠️ Define Weave-Tracked Functions
+
+We'll define our functions for the pipeline and use `@weave.op()` to decorate them, enabling Weave to track their inputs and outputs.
+
+### 1. Retrieve Context
+
+This function retrieves relevant context for the question using the LlamaIndex query engine.
+
+
+```python
+@weave.op()
+def retrieve_context(question: str) -> str:
+    '''
+    Retrieves relevant context for the question using LlamaIndex query engine.
+    '''
+    response = query_engine.query(question)
+    context = str(response)
+    return context
+```
+
+### 2. Generate Answer
+
+This function generates an answer to the question based on the provided context using OpenAI's GPT model.
+
+
+
+```python
+@weave.op()
+def generate_answer(question: str, context: str, model_name: str) -> str:
+    '''
+    Generates an answer to the question based on the provided context using OpenAI's GPT model.
+    '''
+
+    messages = [
+        {"role": "system", "content": "You are a helpful assistant that answers questions based on the provided context."},
+        {"role": "user", "content": f"Context:\n{context}\n\nQuestion:\n{question}"}
+    ]
+    response = client.chat.completions.create(
+        model=model_name,
+        messages=messages,
+        max_tokens=200,
+        temperature=0.7,
+        n=1,
+    )
+    answer = response.choices[0].message.content.strip()
+    return answer
+
+
+```
+
+### 3. Break Down Answer into Statements
+
+This function breaks down the answer into simpler statements without pronouns.
```

Comment:
Why are we doing this?

Reply:
when you get the asnwer from LLM ( lets say in RAG) there are multiple statemtns provided even in one single asnwer so we need to run hallucination evalaution to check whic ones are fact. For instance, you ask about What is weave? Chatbot says: Weave is a product releaed at THIS Date by WANDB. So there are. two statements here and we need to check for both in case of hallucination

Reply:
That's great, we should tell the person reading this that 😉

Reply:
Added to the notebook :) 

_In this step, the response generated by the LLM is divided into multiple individual statements. The reason for breaking down the answer is to evaluate each statement independently for potential hallucinations. Since a single answer can contain a mix of factual and hallucinated information, checking each statement separately ensures a more fine-grained analysis.

This approach helps identify whether every part of the response is grounded in the retrieved context, as some statements might appear factually correct but are not supported by the provided retrieval. This breakdown allows us to better detect factual but ungrounded (hallucinated) content, improving the reliability of the evaluation process._